hypothesis-python/tests/numpy/test_gufunc.py

```diff
@@ -0,0 +1,675 @@
+    elements = sampled_from(choices)
+    S = gu._arrays(dtype, shape, elements)
+    X = data.draw(S)
+
+    assert np.shape(X) == shape
+    validate_elements([X], dtype=dtype, choices=choices)
+
+    assert type(X) == np.ndarray
+
+
+@given(real_scalar_dtypes(), _st_shape, data())
+def test_just_arrays(dtype, shape, data):
+    # unique argument to arrays gets tested in the tuple of arrays tests
+    choices = data.draw(real_from_dtype(dtype))
+
+    # test again, but this time pass in strategy to make sure it can handle it
+    elements = sampled_from(choices)
+    S = gu._arrays(just(dtype), just(shape), elements)
+    X = data.draw(S)
+
+    assert np.shape(X) == shape
+    validate_elements([X], dtype=dtype, choices=choices)
+
+    assert type(X) == np.ndarray
+
+
+# hypothesis.extra.numpy.array_shapes does not support 0 min_size so we roll
+# our own in this case.
+@given(
+    lists(_st_shape, min_size=0, max_size=5), real_scalar_dtypes(), booleans(), data()
+)
+def test_shapes_tuple_of_arrays(shapes, dtype, unique, data):
+    elements = from_dtype(np.dtype(dtype))
+
+    S = gu._tuple_of_arrays(shapes, dtype, elements=elements, unique=unique)
+    X = data.draw(S)
+
+    validate_elements(X, dtype=dtype, unique=unique)
+
+    assert len(shapes) == len(X)
+    for spec, drawn in zip(shapes, X):
+        assert tuple(spec) == np.shape(drawn)
+
+
+# hypothesis.extra.numpy.array_shapes does not support 0 min_size so we roll
+# our own in this case.
+@given(
+    lists(_st_shape, min_size=0, max_size=5), real_scalar_dtypes(), booleans(), data()
+)
+def test_just_shapes_tuple_of_arrays(shapes, dtype, unique, data):
+    elements = from_dtype(np.dtype(dtype))
+
+    # test again, but this time pass in strategy to make sure it can handle it
+    S = gu._tuple_of_arrays(
+        just(shapes), just(dtype), elements=elements, unique=just(unique)
+    )
+    X = data.draw(S)
+
+    validate_elements(X, dtype=dtype, unique=unique)
+
+    assert len(shapes) == len(X)
+    for spec, drawn in zip(shapes, X):
+        assert tuple(spec) == np.shape(drawn)
+
+
+@given(lists(_st_shape, min_size=0, max_size=5), real_scalar_dtypes(), data())
+def test_elements_tuple_of_arrays(shapes, dtype, data):
+    choices = data.draw(real_from_dtype(dtype))
+
+    elements = sampled_from(choices)
+    S = gu._tuple_of_arrays(shapes, dtype, elements=elements)
+    X = data.draw(S)
+
+    validate_elements(X, choices=choices, dtype=dtype)
+
+
+@given(
+    gu.gufunc_args(
+        "(1),(1),(1),()->()",
+        dtype=["object", "object", "object", "bool"],
+        elements=[_st_shape, scalar_dtypes(), just(None), booleans()],
+        min_side=1,
+        max_dims_extra=1,
+    ),
+    data(),
+)
+def test_bcast_tuple_of_arrays(args, data):
+    """Now testing broadcasting of tuple_of_arrays, kind of crazy since it uses
+    gufuncs to test itself. Some awkwardness here since there are a lot of
+    corner cases when dealing with object types in the numpy extension.
```

Comment:
Hopefully fewer of them now, right?  What corner cases, if any, are left?

(the general theme here is that you're pushing the Hypothesis-Numpy integration harder than anyone else has told us about, so we'd love any design and API feedback you might have)

Reply:
It should be better after the last big fix, but there is still some trickiness since numpy will not put  a tuple inside an array with `shape=()` even when `dtype=object`.

Reply:
I can see how that happens, but ahhhh!  We can fix it though:

```python
return np.full(shape=(), fill_value=draw(elements), dtype=dtype)
```

becomes

```python
# We use .itemset() instead of np.full because the latter cannot set tuples as elements.
arr = np.empty(shape=(), dtype=dtype)
arr.itemset(draw(elements))
return arr
```

Reply:
Should we make this change in the same pull request?

Reply:
I'd like to get this and the array_shapes change out separately from the gufuncs change, mainly because there's a serious non-linearity in the work involved in reviewing large PRs.

Ideally that would be a patch release each, but doing one patch release with both would be OK too.  Sorry this is taking so long to get to the main event - but thanks for the stream of already-released improvements!

Reply:
yeah, doing both in the same PR seems more efficient.